File: Python/03/product.py
"""
Moduł demonstracyjny do rozdziału "Klasy i metody" (Python).

Pokazuje:
- atrybuty instancji i klasy,
- docstringi modułu/klasy/metod (PEP 257),
- __init__, @property z walidacją, @classmethod, @staticmethod,
- prostą reprezentację __repr__ i __str__.
"""
from __future__ import annotations

from dataclasses import dataclass


class Product:
    """Reprezentuje produkt w sklepie.

    Atrybuty:
        name: Nazwa produktu (niepusta).
        net_price: Cena netto (>= 0).
        vat_rate: Stawka VAT (0..1). Domyślnie DEFAULT_VAT.
    """

    # Atrybut klasowy (stała wg konwencji):
    DEFAULT_VAT: float = 0.23

    # Prosty licznik instancji (dla przykładu metody @classmethod)
    _COUNTER: int = 0

    def __init__(self, name: str, net_price: float, vat_rate: float | None = None) -> None:
        """Inicjalizuje nowy obiekt Product.

        Args:
            name: Nazwa (niepusta).
            net_price: Cena netto (>= 0).
            vat_rate: Stawka VAT (0..1). Gdy None, używana jest DEFAULT_VAT.
        """
        self.name = name  # przejdzie przez setter z walidacją
        self.net_price = net_price  # setter
        self.vat_rate = self.DEFAULT_VAT if vat_rate is None else vat_rate  # setter
        type(self)._COUNTER += 1

    # Enkapsulacja przez właściwości:
    @property
    def name(self) -> str:
        """Nazwa produktu (niepusta)."""
        return self._name

    @name.setter
    def name(self, value: str) -> None:
        if not value or not value.strip():
            raise ValueError("name is blank")
        self._name = value

    @property
    def net_price(self) -> float:
        """Cena netto (>= 0)."""
        return self._net_price

    @net_price.setter
    def net_price(self, value: float) -> None:
        if value < 0:
            raise ValueError("net_price < 0")
        self._net_price = float(value)

    @property
    def vat_rate(self) -> float:
        """Stawka VAT (0..1)."""
        return self._vat_rate

    @vat_rate.setter
    def vat_rate(self, value: float) -> None:
        if not (0 <= value <= 1):
            raise ValueError("vat_rate out of range")
        self._vat_rate = float(value)

    def gross_price(self) -> float:
        """Zwraca cenę brutto liczona na podstawie bieżących pól instancji."""
        return self.net_price * (1.0 + self.vat_rate)

    @staticmethod
    def gross_of(net: float, vat: float) -> float:
        """Statyczna funkcja pomocnicza do liczenia ceny brutto.

        Args:
            net: Cena netto (>= 0).
            vat: Stawka VAT (0..1).
        """
        if net < 0:
            raise ValueError("net < 0")
        if not (0 <= vat <= 1):
            raise ValueError("vat out of range")
        return float(net) * (1.0 + float(vat))

    @classmethod
    def created_count(cls) -> int:
        """Zwraca liczbę utworzonych instancji Product."""
        return cls._COUNTER

    def __repr__(self) -> str:  # debugowa, nieformalna
        return f"Product(name={self.name!r}, net_price={self.net_price!r}, vat_rate={self.vat_rate!r})"

    def __str__(self) -> str:  # przyjazna
        return f"Product '{self.name}' netto={self.net_price:.2f}, VAT={self.vat_rate:.0%}"


if __name__ == "__main__":
    """Prosty pokaz użycia.

    Uruchom: python3 Python/03/product.py
    """
    apple = Product("Jabłko", 2.50)
    cheese = Product("Ser", 10.0, 0.08)

    print(apple)
    print("Brutto (apple):", apple.gross_price())

    print(cheese)
    print("Brutto (cheese):", cheese.gross_price())

    print("Brutto (100, default VAT):", Product.gross_of(100, Product.DEFAULT_VAT))
    print("Utworzono produktów:", Product.created_count())

File: Java/03/Product.java
/**
 * Reprezentuje produkt w sklepie.
 * <p>
 * Przykład pokazuje:
 * - pola prywatne i statyczne,
 * - konstruktor(y), enkapsulację (gettery/settery z walidacją),
 * - metody instancyjne i statyczne,
 * - Javadoc dla klasy i metod.
 */
public class Product {
    /** Globalny licznik utworzonych produktów. */
    private static int COUNTER = 0;

    /** Domyślna stawka podatku VAT. */
    public static final double DEFAULT_VAT = 0.23;

    /** Unikalny identyfikator produktu. */
    private final int id;

    /** Nazwa produktu (niepusta). */
    private String name;

    /** Cena netto (>= 0). */
    private double netPrice;

    /** Stawka VAT (0..1). */
    private double vatRate;

    /**
     * Tworzy nowy produkt z domyślną stawką VAT.
     *
     * @param name     nazwa produktu (niepusta)
     * @param netPrice cena netto (>= 0)
     * @throws IllegalArgumentException gdy dane są niepoprawne
     */
    public Product(String name, double netPrice) {
        this(name, netPrice, DEFAULT_VAT);
    }

    /**
     * Tworzy nowy produkt z podaną stawką VAT.
     *
     * @param name     nazwa produktu (niepusta)
     * @param netPrice cena netto (>= 0)
     * @param vatRate  stawka VAT (0..1)
     * @throws IllegalArgumentException gdy dane są niepoprawne
     */
    public Product(String name, double netPrice, double vatRate) {
        validateName(name);
        validateNetPrice(netPrice);
        validateVat(vatRate);
        this.id = ++COUNTER;
        this.name = name;
        this.netPrice = netPrice;
        this.vatRate = vatRate;
    }

    /**
     * Zwraca identyfikator produktu.
     * @return id
     */
    public int getId() {
        return id;
    }

    /**
     * Zwraca nazwę produktu.
     * @return nazwa (niepusta)
     */
    public String getName() {
        return name;
    }

    /**
     * Ustawia nazwę produktu.
     * @param name nowa nazwa (niepusta)
     */
    public void setName(String name) {
        validateName(name);
        this.name = name;
    }

    /**
     * Zwraca cenę netto.
     */
    public double getNetPrice() {
        return netPrice;
    }

    /**
     * Ustawia cenę netto.
     * @param netPrice cena (>= 0)
     */
    public void setNetPrice(double netPrice) {
        validateNetPrice(netPrice);
        this.netPrice = netPrice;
    }

    /**
     * Zwraca stawkę VAT.
     */
    public double getVatRate() {
        return vatRate;
    }

    /**
     * Ustawia stawkę VAT.
     * @param vatRate stawka (0..1)
     */
    public void setVatRate(double vatRate) {
        validateVat(vatRate);
        this.vatRate = vatRate;
    }

    /**
     * Oblicza cenę brutto (instancyjnie), stosując aktualne pola obiektu.
     * @return cena brutto
     */
    public double grossPrice() {
        return netPrice * (1.0 + vatRate);
    }

    /**
     * Metoda statyczna obliczająca cenę brutto na podstawie argumentów.
     * @param net cena netto
     * @param vat stawka VAT
     * @return cena brutto
     */
    public static double grossOf(double net, double vat) {
        if (net < 0) throw new IllegalArgumentException("net < 0");
        if (vat < 0 || vat > 1) throw new IllegalArgumentException("vat out of range");
        return net * (1.0 + vat);
    }

    /**
     * Zwraca liczbę utworzonych produktów.
     */
    public static int createdCount() {
        return COUNTER;
    }

    private static void validateName(String name) {
        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("name is blank");
        }
    }

    private static void validateNetPrice(double price) {
        if (price < 0) {
            throw new IllegalArgumentException("price < 0");
        }
    }

    private static void validateVat(double vat) {
        if (vat < 0 || vat > 1) {
            throw new IllegalArgumentException("vat out of range");
        }
    }

    /**
     * Reprezentacja tekstowa produktu.
     */
    @Override
    public String toString() {
        return "Product{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", netPrice=" + netPrice +
                ", vatRate=" + vatRate +
                '}';
    }
}
